conda/common/configuration.py

```diff
@@ -1609,3 +1609,54 @@ def typify_parameter(self, parameter_name, value, source):
 
     def get_descriptions(self):
         raise NotImplementedError()
+
+
+def get_plugin_config_data(
+    data: dict[Path, dict[str, RawParameter]],
+) -> dict[Path, dict[str, RawParameter]]:
+    """
+    This is used to move everything under the key "plugins" from the provided dictionary
+    to the top level of the returned dictionary. The returned dictionary is then passed
+    to :class:`PluginConfig`.
+    """
+    new_data = {}
+
+    for source, config in data.items():
+        if plugin_data := config.get("plugins"):
+            for param_name, raw_param in plugin_data.value(None).items():
+                new_data[source] = {param_name: raw_param}
+
+        if source == "envvars":
+            for env_var, raw_param in config.items():
+                if env_var.startswith("plugins_"):
+                    _, param_name = env_var.split("plugins_")
+                    new_data[source] = {param_name: raw_param}
+
+    return new_data
+
+
+class PluginConfigBase(metaclass=ConfigurationType):
+    """Base class for creating subclassing new types from"""
+
+
+def create_plugin_config_type(plugin_settings: dict[str, ParameterLoader]) -> type:
+    """Creates a type used to hold plugin configuration settings"""
+
+    def __init__(self, data):
+        self._cache_ = {}
+        self.raw_data = get_plugin_config_data(data)
+
+    plugin_config_type = type(
+        "PluginConfig",
+        (PluginConfigBase,),
+        {
+            "__init__": __init__,
+        },
+    )
+
+    for name, loader in plugin_settings.items():
+        plugin_config_type.parameter_names += (name,)
+        name = loader._set_name(name)
+        setattr(plugin_config_type, name, loader)
+
+    return plugin_config_type
```

Comment:
Please no meta programming like this, it's going to make things much harder later, it's one of those cases where being clever isn't worth it